src/mtev_cluster.c

```diff
@@ -223,35 +253,80 @@ mtev_cluster_check_timeouts() {
 
 static int 
 mtev_cluster_info_compose(void *payload, int len, void *c) {
+  hb_payload_t *hb_payload;
+
   mtev_cluster_check_timeouts();
 
   MEMWRITE_DECL(payload, len);
   u_int64_t packed_time;
-  u_int64_t seq;
+  u_int64_t my_seq;
   mtev_cluster_t *cluster = c;
+  mtev_hash_iter payload_iter = MTEV_HASH_ITER_ZERO;
+  mtev_hash_iter payload_iter2 = MTEV_HASH_ITER_ZERO;
+  void *vdata;
+  const char *key;
+  int klen;
+
   /* payload about the local cluster goes here */
   /* The header is always the same: <uuid:16><boottime_sec:5><boottiime_usec:3><seq:8> */
   MEMWRITE(my_cluster_id, UUID_SIZE);
   packed_time = (my_boot_time.tv_sec & 0x000000ffffffffffULL) << 24; /* 5 bytes << 3 bytes */
   packed_time |= (my_boot_time.tv_usec & 0xffffff); /* 3 bytes */
   packed_time = htonll(packed_time);
   MEMWRITE(&packed_time, sizeof(packed_time));
-  seq = htonll(cluster->config_seq);
-  MEMWRITE(&seq, sizeof(seq));
+  my_seq = htonll(cluster->config_seq);
+  MEMWRITE(&my_seq, sizeof(my_seq));
+
+  assert(HEART_BEAT_HDR_LEN == MEMWRITE_WRITTEN);
+
+  if(mtev_hash_size(&cluster->hb_payloads) != 0) {
+    // write header (pointer table)
+    while(mtev_hash_next(&cluster->hb_payloads, &payload_iter, &key, &klen, &vdata)) {
+      hb_payload = vdata;
+      MEMWRITE(&hb_payload->app_id, 1);
+      MEMWRITE(&hb_payload->key, 1);
+      MEMWRITE(&hb_payload->data_len, 1);
+    }
+    // write payload
+    while(mtev_hash_next(&cluster->hb_payloads, &payload_iter2, &key, &klen, &vdata)) {
+      hb_payload = vdata;
+      MEMWRITE(hb_payload->data, hb_payload->data_len);
```

Comment:
That's actually very typical for my experience. It's a similar concept to the common rule for high performance: Use structs of arrays instead of arrays of structs to reduce cash misses.

The concept here is that you have a very compact list of headers fitting into one page. You can quickly iterate over it to search for the header you're looking for while summing up the payload lengths. Once your sum is larger than the total size of your data block you know that you've reached the end of the header list. Otherwise the sum points to the payload. This is done here:  https://github.com/JonasKunze/libmtev/blob/smart_cluster/src/mtev_cluster.c#L815

You can argue that a protocol with a high cash locality is not important here as we're anyway not focusing on performance, but that's simply how I've learned to design binary protocols many times.

If you want, I can change this implementation to what you've suggested.